File: scripts/audit_licenses.py
"""Read the licence of every Python distribution the environment installs.

The counterpart of ``web/scripts/audit-licenses.mjs``, reading the same
``license-policy.json``: a licence decision is made once for the repository
rather than once per language.

It is an **engineering gate, not legal advice**. It answers one question -- "is
every licence here one this repository has already decided it accepts?" -- and
reports everything else for a human to read rather than claiming a verdict it is
not qualified to give.

```
allowed          on the repository's allowlist, or reviewed under this very licence
review required  a real licence, not on the list
relicensed       reviewed once, and no longer under the licence that was read
unknown          nothing stated at all
```

``relicensed`` exists because a manual exception is a statement about a
*licence*. Recorded against a package name alone it would survive that package
being relicensed, which is the one moment somebody needs to look again.

Metadata is read from the *installed* distributions rather than from
``pyproject.toml``, because what ends up in a container is what was resolved.
Older packaging metadata states a licence in a classifier instead of an SPDX
expression, so both are read and the expression wins where there is one.
"""

from __future__ import annotations

import json
import sys
from collections.abc import Mapping
from dataclasses import dataclass
from importlib.metadata import distributions
from pathlib import Path
from typing import Any

POLICY_FILE = Path(__file__).resolve().parents[1] / "license-policy.json"

CLASSIFIER_PREFIX = "License :: "
"""Where a distribution that predates SPDX metadata states its licence."""

_CLASSIFIER_SPDX = {
    "MIT License": "MIT",
    "Apache Software License": "Apache-2.0",
    "BSD License": "BSD-3-Clause",
    "ISC License (ISCL)": "ISC",
    "Mozilla Public License 2.0 (MPL 2.0)": "MPL-2.0",
    "Python Software Foundation License": "Python-2.0",
    "The Unlicense (Unlicense)": "Unlicense",
}
"""Classifiers this project has actually met, mapped onto SPDX identifiers.

Deliberately small and deliberately not a general translation table. A
classifier is coarser than an expression -- "BSD License" names a family, not a
document -- so anything not listed here is reported as needing a look rather
than guessed at.
"""


@dataclass(frozen=True, slots=True)
class Package:
    """One installed distribution and what it says its licence is."""

    name: str
    version: str
    license: str | None


def installed_packages() -> list[Package]:
    """Return every installed distribution, sorted by name."""
    found: dict[str, Package] = {}
    for distribution in distributions():
        metadata = distribution.metadata
        name = metadata["Name"]
        if not name:
            continue
        found[name.lower()] = Package(
            name=name,
            version=distribution.version or "?",
            license=_declared(metadata),
        )
    return sorted(found.values(), key=lambda package: package.name.lower())


def _declared(metadata: object) -> str | None:
    """Return the SPDX expression a distribution states, if it states one."""
    get = getattr(metadata, "get", None)
    get_all = getattr(metadata, "get_all", None)
    if get is None or get_all is None:  # pragma: no cover - defensive
        return None
    expression = get("License-Expression")
    if isinstance(expression, str) and expression.strip():
        return expression.strip()
    for classifier in get_all("Classifier") or []:
        if not isinstance(classifier, str) or not classifier.startswith(CLASSIFIER_PREFIX):
            continue
        spdx = _CLASSIFIER_SPDX.get(classifier.rsplit(" :: ", 1)[-1])
        if spdx:
            return spdx
    legacy = get("License")
    if isinstance(legacy, str) and legacy.strip() and "\n" not in legacy.strip():
        return legacy.strip()
    return None


def verdict(expression: str | None, allowed: frozenset[str]) -> str:
    """Return what one licence expression amounts to under the policy.

    ``OR`` passes when either side does: a package offered under a choice may be
    taken under the one this repository accepts. ``AND`` requires every part,
    because both apply.
    """
    if not expression:
        return "unknown"
    cleaned = expression.replace("(", " ").replace(")", " ").strip()
    if " AND " in cleaned.upper():
        parts = [part.strip() for part in _split(cleaned, " AND ")]
        return "allowed" if all(part in allowed for part in parts) else "review"
    parts = [part.strip() for part in _split(cleaned, " OR ")]
    return "allowed" if any(part in allowed for part in parts) else "review"


def reviewed_licenses(policy: Mapping[str, Any]) -> dict[str, str]:
    """Return the licence each individually reviewed package was accepted under.

    Names are lowered because a distribution name is not case sensitive, and an
    exception that only applies to one spelling is an exception that stops
    applying when somebody writes the other one.
    """
    return {name.lower(): entry["license"] for name, entry in policy.get("reviewed", {}).items()}


def decide(package: Package, allowed: frozenset[str], reviewed: Mapping[str, str]) -> str:
    """Return what the policy makes of one package.

    A reviewed package is allowed **under the licence it was reviewed under**
    and under no other. That is the whole point of recording the licence beside
    the reason: the exception describes a document somebody read, so a package
    that now arrives under a different one has not been reviewed at all.
    """
    expected = reviewed.get(package.name.lower())
    if expected is None:
        return verdict(package.license, allowed)
    if package.license == expected:
        return "allowed"
    return "relicensed"


def _split(expression: str, separator: str) -> list[str]:
    """Split on a separator, ignoring case."""
    lowered = expression.upper()
    parts: list[str] = []
    start = 0
    while True:
        found = lowered.find(separator, start)
        if found < 0:
            parts.append(expression[start:])
            return parts
        parts.append(expression[start:found])
        start = found + len(separator)


def main() -> int:
    """Report the installed licences and fail on anything unreviewed."""
    policy = json.loads(POLICY_FILE.read_text(encoding="utf-8"))
    allowed = frozenset(policy["allowed"])
    reviewed = reviewed_licenses(policy)
    # This repository's own distributions are not third-party dependencies, and
    # auditing a project against a policy it wrote says nothing.
    own = {name.lower().replace("-", "_") for name in policy.get("self", [])}

    packages = [
        package
        for package in installed_packages()
        if package.name.lower().replace("-", "_") not in own
    ]
    counts: dict[str, int] = {}
    problems: list[tuple[Package, str]] = []
    for package in packages:
        counts[package.license or "unstated"] = counts.get(package.license or "unstated", 0) + 1
        decided = decide(package, allowed, reviewed)
        if decided != "allowed":
            problems.append((package, decided))

    lines = [f"{len(packages)} installed distributions"]
    for license_name, count in sorted(counts.items(), key=lambda item: (-item[1], item[0])):
        lines.append(f"  {count:>4}  {license_name}")
    if problems:
        lines.append("")
        lines.append("Not covered by the repository policy:")
        for package, decided in problems:
            stated = package.license or "unstated"
            note = (
                f" (reviewed as {reviewed[package.name.lower()]})"
                if decided == "relicensed"
                else ""
            )
            lines.append(f"  {decided:<10} {package.name}=={package.version} — {stated}{note}")
        lines.append("")
        lines.append("Add the identifier to license-policy.json only after reading the licence.")
    else:
        lines.append("")
        lines.append("Every installed distribution carries a licence this repository accepts.")

    sys.stdout.write("\n".join(lines) + "\n")
    return 1 if problems else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: web/scripts/audit-licenses.mjs
#!/usr/bin/env node
/**
 * Read the licence of every package `npm ci` actually installs.
 *
 * `package.json` states intent; `package-lock.json` states what is resolved,
 * and it is the resolution that ends up in a container. So the tree is walked
 * from `node_modules`, where the licence text and the package metadata that
 * names it actually live.
 *
 * This is an **engineering gate, not legal advice**. It answers one question --
 * "is every licence here one this repository has already decided it accepts?"
 * -- and it reports everything else for a human to look at rather than claiming
 * a verdict it is not qualified to give.
 *
 *   allowed          on the allowlist, or reviewed under this very licence
 *   review required  a real licence, not on the list
 *   relicensed       reviewed once, and no longer under the licence that was read
 *   unknown          no licence stated at all
 *
 * `relicensed` exists because a manual exception is a statement about a
 * *licence*. Recorded against a package name alone it would survive that
 * package being relicensed, which is the one moment somebody needs to look
 * again.
 *
 * Exits non-zero when anything is not `allowed`, so a dependency that arrives
 * with a surprise is noticed when it arrives rather than at a release.
 */
import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs'
import { join } from 'node:path'
import { fileURLToPath } from 'node:url'

const ROOT = fileURLToPath(new URL('..', import.meta.url))
const MODULES = join(ROOT, 'node_modules')
// One policy for the whole repository: the Python audit reads the same file,
// so a licence decision is made once rather than once per language.
const POLICY = JSON.parse(readFileSync(join(ROOT, '..', 'license-policy.json'), 'utf8'))

const ALLOWED = new Set(POLICY.allowed)
// A reviewed package is accepted under the licence somebody read, and under no
// other. The value is the expected identifier, not a free-text note.
const REVIEWED = new Map(
  Object.entries(POLICY.reviewed ?? {}).map(([name, entry]) => [name, entry.license]),
)
// This repository's own packages are not third-party dependencies.
const OWN = new Set(POLICY.self ?? [])

/** Every installed package, by name, with the version and licence it declares. */
function installed() {
  const found = new Map()
  const walk = (directory, scope = null) => {
    if (!existsSync(directory)) return
    for (const entry of readdirSync(directory)) {
      if (entry === '.bin' || entry === '.package-lock.json') continue
      const path = join(directory, entry)
      if (!statSync(path).isDirectory()) continue
      if (entry.startsWith('@') && scope === null) {
        walk(path, entry)
        continue
      }
      const name = scope === null ? entry : `${scope}/${entry}`
      const manifest = join(path, 'package.json')
      if (existsSync(manifest)) {
        const meta = JSON.parse(readFileSync(manifest, 'utf8'))
        found.set(name, { version: meta.version ?? '?', license: declared(meta) })
      }
      walk(join(path, 'node_modules'))
    }
  }
  walk(MODULES)
  return found
}

/** What a package's own metadata says its licence is. */
function declared(meta) {
  if (typeof meta.license === 'string') return meta.license
  if (meta.license && typeof meta.license.type === 'string') return meta.license.type
  if (Array.isArray(meta.licenses) && meta.licenses.length > 0) {
    return meta.licenses.map((entry) => entry.type ?? '?').join(' OR ')
  }
  return null
}

/**
 * Split an SPDX expression into the identifiers it names.
 *
 * `(MIT OR Apache-2.0)` is allowed when either side is: a package offered under
 * a choice may be taken under the one this repository accepts. `AND` is
 * different -- both apply -- so every part has to pass.
 */
function verdict(expression) {
  if (!expression) return 'unknown'
  const cleaned = expression.replace(/[()]/g, ' ').trim()
  if (/\bAND\b/i.test(cleaned)) {
    const parts = cleaned.split(/\s+AND\s+/i).map((part) => part.trim())
    return parts.every((part) => ALLOWED.has(part)) ? 'allowed' : 'review'
  }
  const parts = cleaned.split(/\s+OR\s+/i).map((part) => part.trim())
  return parts.some((part) => ALLOWED.has(part)) ? 'allowed' : 'review'
}

const packages = [...installed().entries()]
  .filter(([name]) => !OWN.has(name))
  .sort(([a], [b]) => a.localeCompare(b))
const problems = []
const counts = new Map()

/**
 * What the policy makes of one package.
 *
 * A reviewed package is allowed under the licence it was reviewed under and
 * under no other: the exception describes a document somebody read, so the same
 * package arriving under a different licence has not been reviewed at all.
 */
function decide(name, license) {
  if (!REVIEWED.has(name)) return verdict(license)
  return license === REVIEWED.get(name) ? 'allowed' : 'relicensed'
}

for (const [name, { version, license }] of packages) {
  const decided = decide(name, license)
  counts.set(license ?? 'unstated', (counts.get(license ?? 'unstated') ?? 0) + 1)
  if (decided !== 'allowed') {
    problems.push({
      name,
      version,
      license: license ?? 'unstated',
      decided,
      reviewed: REVIEWED.get(name) ?? null,
    })
  }
}

process.stdout.write(`${packages.length} installed packages\n`)
for (const [license, count] of [...counts.entries()].sort((a, b) => b[1] - a[1])) {
  process.stdout.write(`  ${String(count).padStart(4)}  ${license}\n`)
}

if (problems.length > 0) {
  process.stdout.write('\nNot covered by the repository policy:\n')
  for (const problem of problems) {
    const note = problem.decided === 'relicensed' ? ` (reviewed as ${problem.reviewed})` : ''
    process.stdout.write(
      `  ${problem.decided.padEnd(10)} ${problem.name}@${problem.version} — ${problem.license}${note}\n`,
    )
  }
  process.stdout.write(
    '\nAdd the identifier to license-policy.json only after somebody has read the licence.\n',
  )
  process.exit(1)
}

process.stdout.write('\nEvery installed package carries a licence this repository accepts.\n')
